lib/src/utils/serialization_utils.dart

```diff
@@ -340,4 +363,386 @@ class SerializationUtils {
+            serializeOptional(writer, relay.dnsName, (w, value) => w.write(serializeDnsName(value)));
+          },
+      };
+      invoker();
+      return writer.toBytes();
+    });
+  }
+
+  static void serializeOptional<T>(
+      ByteDataWriter writer, T? value, void Function(ByteDataWriter, T) serializeFunction) {
+    if (value == null) {
+      writer.writeUint8(1);
+    } else {
+      writer.writeUint8(2);
+      serializeFunction(writer, value);
+    }
+  }
+
+  static Uint8List serializeIpv4(String ipv4) {
+    return ipStringToBytes(ipv4);
+  }
+
+  static Uint8List serializeIpv6(String ipv6) {
+    return ipStringToBytes(ipv6);
+  }
+
+  static Uint8List serializeDnsName(String dnsName) {
+    return Uint8List.fromList(utf8.encode(dnsName));
+  }
+
+  static Uint8List serializePoolMetadata(ParsedPoolMetadata? metadata) {
+    return useBinaryWriter((ByteDataWriter writer) {
+      if (metadata == null) {
+        serializeOptionFlag(writer, false);
+      } else {
+        serializeOptionFlag(writer, true);
+        writeSerializedHex(writer, metadata.hashHex);
+        writer.write(Uint8List.fromList(utf8.encode(metadata.url)));
+      }
+      return writer.toBytes();
+    });
+  }
+
+  static Uint8List serializeTxWithdrawal(ParsedWithdrawal withdrawal, CardanoVersion version) {
+    if (VersionCompatibility.checkVersionCompatibility(version).supportsMultisigTransaction) {
+      return useBinaryWriter((ByteDataWriter writer) {
+        writer.write(serializeCoin(withdrawal.amount));
+        writer.write(serializeCredential(withdrawal.stakeCredential));
+        return writer.toBytes();
+      });
+    } else {
+      final withdrawalStakeCredential = withdrawal.stakeCredential;
+      if (withdrawalStakeCredential is! CredentialKeyPath) {
+        throw ValidationException('WITHDRAWAL_INVALID_STAKE_CREDENTIAL');
+      }
+      return useBinaryWriter((ByteDataWriter writer) {
+        writer.write(serializeCoin(withdrawal.amount));
+        writerSerializedPath(writer, withdrawalStakeCredential.path);
+        return writer.toBytes();
+      });
+    }
+  }
+
+  static Uint8List serializeTxValidityStart(String validityIntervalStart) {
+    return useBinaryWriter((ByteDataWriter writer) {
+      writeSerializedUint64(writer, BigInt.parse(validityIntervalStart));
+      return writer.toBytes();
+    });
+  }
+
+  static Uint8List serializeMintBasicParams(List<ParsedAssetGroup> mint) {
+    return useBinaryWriter((ByteDataWriter writer) {
+      writer.writeUint32(mint.length);
+      return writer.toBytes();
+    });
+  }
+
+  static Uint8List serializeAssetGroup(ParsedAssetGroup assetGroup) {
+    return useBinaryWriter((ByteDataWriter writer) {
+      writeSerializedHex(writer, assetGroup.policyIdHex);
+      writer.writeUint32(assetGroup.tokens.length);
+      return writer.toBytes();
+    });
+  }
+
+  static Uint8List serializeToken<T>(ParsedToken token, Uint8List Function(T) serializeTokenAmountFn) {
```

Comment:
I am a bit confused why `serializeTokenAmountFn` is passed (and why it is  a function). Maybe I am wrong but I expect this function should always be the same, which means it shouldn't be passed as argument.

Reply:
clarified it in the latest commit